### ISSUE TITLE
Wheel support?

### ISSUE THREAD

**Issue by Varbin** [jonashaag/bjoern#121]
Having wheels on PyPi could make installation a lot faster and may also solve the "library headers are required for installation"-problem.

**Comment by Varbin** [comment]
Some CI services could be used. On Travis OS X containers can be used (but I have no experience with them) and AppVeyor  (See https://packaging.python.org/guides/supporting-windows-using-appveyor/) makes it possible  to build on Windows - still this all has to be set up.

For Linux an "manylinux" wheel would be great, it is possible to also build them on travis (example project: https://github.com/pypa/python-manylinux-demo).

**Comment by jonashaag** [comment]
Ok, I'm happy to incorporate this into the release process as long as it means I don't have to install Linux or Windows. Even better if this could be done completely on CI or some trustworthy third party service.

**Comment by jonashaag** [comment]
(Docker is fine as well)

**Comment by Varbin** [comment]
@jonashaag The manylinux wheel creation is done with Docker.

**Comment by jonashaag** [comment]
I see; I'll have a look

**Comment by jonashaag** [comment]
Looks good but seems to be a bit of work. Feel free to contribute a working PR :)

**Comment by Varbin** [comment]
As Continuous Integration could also run a test suite, would you mind if I convert some of the present tests into some independently running automated tests? So before building wheels, it will be tested if nothing broke.

**Comment by Varbin** [comment]
Before doing so, I would like to finish #122/#123 first.

**Comment by jonashaag** [comment]
> As Continuous Integration could also run a test suite, would you mind if I convert some of the present tests into some independently running automated tests? So before building wheels, it will be tested if nothing broke.

That would be absolutely awesome. See also #52. I never got around to doing it.

**Comment by Jitsusama** [comment]
I would like to say that I find this PR quite interesting. Based upon corporate restraints, I have to develop on a Windows machine, even though I deploy to Linux. It would be nice to be able to have bjoern run on my development environment.

**Comment by dgilman** [comment]
I have binary wheels built for several Debian and Ubuntu releases here https://github.com/dgilman/bjoern/releases

**Comment by bersace** [comment]
Does the wheel embed libev or do you still need to install libev4 ? does manylinux requires plateforme to ship libev ?

**Comment by hoefling** [comment]
Here's my humble suggestion: a Travis job configuration that builds fat Linux/MacOS wheels with `libev` included, installs them and runs unit tests, with additional releasing on tag push. Examples:

* [Standard build, triggered on branch push](https://travis-ci.org/hoefling/bjoern/builds/564489401)
* [Release, triggered on tag push](https://travis-ci.org/hoefling/bjoern/builds/564489621)

This is how the release would look like (a fake 3.0.3 in this example):

* [PyPI](https://test.pypi.org/project/bjoern/3.0.3/)
* [Github](https://github.com/hoefling/bjoern/releases/tag/3.0.3)

I also started to port the scripts from the `tests/` directory to unit tests (sample tests already executed in jobs), will drop some thoughts in #52. In preparation for that, I configured collecting code coverage and uploading it to `codecov`: [here's what it can look like in the end](https://codecov.io/gh/hoefling/bjoern).

**Comment by bersace** [comment]
Actually, releasing one plateform at a time may help make a step further.

**Comment by bersace** [comment]
Just a note, we should ship a variant with statd enabled. I thinks of another wheel named `bjoern-statd`.

**Comment by bersace** [comment]
@jonashaag what is the release process as of now ? Do you push manually artefacts to PyPI ?

**Comment by bersace** [comment]
Ok, I found `make upload`. https://github.com/jonashaag/bjoern/blob/master/Makefile#L116-L117 Is it relevant ?

**Comment by jonashaag** [comment]
Yes that's what I'm using. And that's the entire release process for now.

I'm happy to change to any other release process ;)

**Comment by pythonmobile** [comment]
Any update on wheels ? I would like to not have to install libev using apt for pip installing bjoern. Thanks.

**Comment by danielsamuels** [comment]
For what it's worth, I wanted to build Bjoern wheels for an internal use-case so I wrote a Dockerfile to do it:

```dockerfile
FROM quay.io/pypa/manylinux2014_x86_64
RUN yum install -y gcc libev-devel \
    && /opt/python/cp38-cp38/bin/pip3.8 wheel bjoern==3.1.0 \
    && auditwheel repair /bjoern-3.1.0-cp38-cp38-linux_x86_64.whl
```

This generates `/wheelhouse/bjoern-3.1.0-cp38-cp38-manylinux2014_x86_64.whl` which can then be installed on a system which does not have libev installed.

**Comment by bersace** [comment]
#178  added plumbing for wheels on Linux. But @jonashaag didn't published wheels yet on PyPI. Today, I also had a look at https://cibuildwheel.readthedocs.io/en/stable/setup/#github-actions which may improve the situation.
